ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.engine.KsqlEngine;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.confluent.ksql.util.QueryMetadata;
+import java.net.URI;
+import java.net.URL;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.state.HostInfo;
+import org.apache.kafka.streams.state.StreamsMetadata;
+
+public final class HeartbeatHandler {
+
+  private final KsqlEngine engine;
+  private final ServiceContext serviceContext;
+  private final ConfigurationParameters configs;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
+  private final ConcurrentHashMap<String, Boolean> aliveHosts;
+  private final ConcurrentHashMap<URI, Boolean> remoteHostsURI;
+  private final ConcurrentHashMap<HostInfo, Boolean> remoteHostsInfo;
```

Comment:
I'm not sure that making these each individually `Concurrent` is sufficient for keeping all of them in sync. It might be good to have one map that maintains all of the information for a given remote host. It makes synchronization easier to reason about (also would be made easier by my suggestion above to use a single thread)

Reply:
+1   I think I left some of this in your fork's PR as well. Its also hard to reason across the maps. Let's have a `HostStatus` or some small struct to manage this more easily.